File: poller.php
#!/usr/bin/env php
<?php
/*
Coding: utf-8
=============================================================================
Author: Christopher Tyler
Email: chris@totalhighspeed.net
Date created: 21JUN2021
Date last modified: 21JUN2021
PHP Version: 7.3
Program Version: 0.0.0
=============================================================================
Pinouts:
PCB v2.2
GPIO  BCM PIN DESCRIPTION
25    D26	37	Contact 1
27    D16	36	Contact 2
24    D19	35	Contact 3
23    D13	33	Contact 4
26    D12	32	Contact 5
22    D6	31	Contact 6

PCB v3.0
GPIO  PIN	DESCRIPTION
4     16	Relay 1
3     15	Relay 2
27    36	Contact 1
0     11	Contact 2
1     12	Contact 3
24    35	Contact 4
28    38	Contact 5
29    40	Contact 6
22    31	PWR LED
=============================================================================
NOTE: For whatever reason, GPIO pin 22 (Contact 6) is borked on the Raspberry Pi 4 and
will not read correctly, even if the internal resistor is set to pulldown mode. If not
set to pulldown mode, the pin always reads as high, and if set to pulldown it will
always read low, regarless if the pin is connected across the 3v rail or not.
I have not tried the PCB v3.0 pinouts yet so I don't know if they will work better or
not. Pin 22 is set as a relay output on v3.0 boards and I think that will work as one
would expect. However, now the other pins are questionable now so we'll see. Might have
to make one more hardware revision to get everything working properly.

TODO: Need to figure out how to send a recovery notification as well, right now it just stops sending emails
but there is no notification when a contact ceases to be in an alarm state.
*/
exec("gpio mode 29 out&&gpio write 29 off");
//Import PHPMailer classes into the global namespace
//These must be at the top of your script, not inside a function
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;
date_default_timezone_set("America/Chicago");

//Load Composer's autoloader
require 'vendor/autoload.php';

function storealarm($counter,$contact_alarm){
  # Store the data only if there isn't a time stored already
  if($contact_alarm == NULL){
    print "Alarm data stored in database\r\n";
    $contact_name = 'contact'.$counter.'_alarm';
    $pdo = new PDO('sqlite:/home/pi/AlarmPiHat/ramdisk/config.db');
    $stm = $pdo->query("UPDATE config SET $contact_name=datetime('now','localtime') WHERE 1");
    $stm->execute();
  }
}

function clearalarm($counter){
  // If there isn't currently and alarm then clear the contact(x)_alarm field in the database
  $contact_name = 'contact'.$counter.'_alarm';
  $notification_name = 'notification'.$counter;
  $pdo = new PDO('sqlite:/home/pi/AlarmPiHat/ramdisk/config.db');
  $stm = $pdo->query("UPDATE config SET $contact_name=NULL, $notification_name=NULL WHERE 1");
  $stm->execute();
}

function checkalarm($contact,$notification_timer){
  // Get the mail and contact naming information from the database
  $pdo = new PDO('sqlite:/home/pi/AlarmPiHat/ramdisk/config.db');
  $stm = $pdo->query("SELECT * FROM config");
  $stm->execute();
  $rows = $stm->fetchAll(PDO::FETCH_ASSOC);
  foreach($rows as $row){
    $notification1 = $row['notification1'];
    $notification2 = $row['notification2'];
    $notification3 = $row['notification3'];
    $notification4 = $row['notification4'];
    $notification5 = $row['notification5'];
    $notification6 = $row['notification6'];
    $notificaiton_timer = $row['alert_timer'];
  }

  $alarm_time = $row['notification'.$contact];
  $alarm_start = new datetime($alarm_time);
  $rightnow = new datetime();
  $duration = $alarm_start->diff($rightnow);
  print "Notificaiton Contact ".$contact.": ".$duration->d." Days ".$duration->h." Hours ".$duration->i." Minutes\r\n";

  if(!empty($alarm_time)){
    if($duration->i >= $notification_timer){
      print "$notification_timer minutes or more have passed, sending message\r\n";
      $status = "send";
    } else {
      print "Not sending message, $notification_timer minutes have not passed since last alarm\r\n";
      $status = "no_send";
    }
  } else {
    print "No previous alarm state, sending message\r\n";
    $status = "send";
  }
  return $status;
}

function mailer($contact,$now,$notification_timer) {
  // Check to see if last notification was within the timer or not
  $alarm = checkalarm($contact,$notification_timer);

  // Get the mail and contact naming information from the database
  $pdo = new PDO('sqlite:/home/pi/AlarmPiHat/ramdisk/config.db');
  $stm = $pdo->query("SELECT * FROM config");
  $stm->execute();
  $rows = $stm->fetchAll(PDO::FETCH_ASSOC);
  foreach($rows as $row){
    $contact_1 = $row['contact_name_1'];
    $contact_2 = $row['contact_name_2'];
    $contact_3 = $row['contact_name_3'];
    $contact_4 = $row['contact_name_4'];
    $contact_5 = $row['contact_name_5'];
    $contact_6 = $row['contact_name_6'];
    $contact1_alarm = $row['contact1_alarm'];
    $contact2_alarm = $row['contact2_alarm'];
    $contact3_alarm = $row['contact3_alarm'];
    $contact4_alarm = $row['contact4_alarm'];
    $contact5_alarm = $row['contact5_alarm'];
    $contact6_alarm = $row['contact6_alarm'];
    $notification1 = $row['notification1'];
    $notification2 = $row['notification2'];
    $notification3 = $row['notification3'];
    $notification4 = $row['notification4'];
    $notification5 = $row['notification5'];
    $notification6 = $row['notification6'];
    $mailto = $row['email_to'];
    $mailfrom = $row['email_from'];
    $subject = $row['email_subject'];
    $smtp_server = $row['smtp_server'];
    $smtp_user = $row['smtp_user'];
    $smtp_pass = $row['smtp_pass'];
  }

  // Corrolate the name to the contact in alarm state
  $contact_name = $row['contact_name_'.$contact];
  $contact_alarm = $row['contact'.$contact.'_alarm'];
  $notification = $row['notification'.$contact];

  if($alarm == "send"){
    // Diff the time so we know how long the alarm has been active
    $alarm_start = new datetime($contact_alarm);
    $rightnow = new datetime();
    $duration = $alarm_start->diff($rightnow);

    //Create an instance; passing `true` enables exceptions
    $mail = new PHPMailer(true);
    try {
      // Format the message
      $msg  = "Date     : ".$now->format("Y-M-d")."\r\n";
      $msg .= "Time     : ".$now->format("H:i:s")."\r\n";
      $msg .= "Contact  : ".$contact_name."\r\n";
      $msg .= "Duration : ".$duration->d." Days ".$duration->h." Hours ".$duration->i." Minutes\r\n";

      // Subject line
      $subject = $subject." - ".$contact_name." detected";

      //Server settings
      //$mail->SMTPDebug = SMTP::DEBUG_SERVER;                      //Enable verbose debug output
      $mail->isSMTP();                                            //Send using SMTP
      $mail->Host       = $smtp_server;                           //Set the SMTP server to send through
      $mail->SMTPAuth   = true;                                   //Enable SMTP authentication
      $mail->Username   = $smtp_user;                             //SMTP username
      $mail->Password   = $smtp_pass;                             //SMTP password
      $mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS;            //Enable implicit TLS encryption
      $mail->Port       = 465;                                    //TCP port to connect to; use 587 if you have set `SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

      //Recipients
      $mail->setFrom($mailfrom, 'Alarm Monitor');
      //$mail->addAddress($mailto, 'User');                         //Add a recipient
      $mail->addAddress('chris@totalhighspeed.net','Chris Tyler');
      $mail->addReplyTo('noreply@totalhighspeed.net', 'NoReply');

      //Content
      $mail->isHTML(false);                                       //Set email format to plaintext
      $mail->Subject = $subject;
      $mail->Body    = $msg;

      $mail->send();
      echo "Message has been sent\r\n";

      // Write $notification to the database
      $notification_name = "notification$contact";
      $pdo = new PDO('sqlite:/home/pi/AlarmPiHat/ramdisk/config.db');
      $stm = $pdo->query("UPDATE config SET $notification_name=datetime('now','localtime') WHERE 1");
      $stm->execute();
    } catch (Exception $e) {
      echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}\r\n";
    }
  }


  return $contact_alarm;
}

$now = new DateTime;

// Get status of contacts
$con1 = exec("gpio read 25");
$con2 = exec("gpio read 27");
$con3 = exec("gpio read 24");
$con4 = exec("gpio read 23");
$con5 = exec("gpio read 26");
$con6 = exec("gpio read 22");

$contacts=array($con1,$con2,$con3,$con4,$con5,$con6);
$counter = 0;
foreach($contacts as $contact){
  $counter++;
  if($contact == 1){
    $contact_alarm = mailer($counter,$now,$notification_timer);
    storealarm($counter,$contact_alarm);
  } else {
    clearalarm($counter);
  }
}
exec("gpio mode 29 out&&gpio write 29 on");
?>

File: poller.py
#!/usr/bin/env python3
# Coding: utf-8
# =============================================================================
# Author: Christopher Tyler
# Email: chris@totalhighspeed.net
# Date created: 05AUG2021
# Python Version: 3
# Program Version: 0.1.0
# =============================================================================
# Documentation:
# This is for AlarmPiHat hardware v3.0
# =============================================================================
# Imports:
import wiringpi
from time import sleep
import sqlite3
import smtplib, ssl
import board
import adafruit_am2320

# =============================================================================
# Codebase:
# Set the IO function for wiringPi to use the BCM pinout
wiringpi.wiringPiSetupGpio()

def getData():
    # Initialize the result variable as a dictionary
    result = {"contact1_alarm":0,"contact2_alarm":0,"contact3_alarm":0,"contact4_alarm":0,"contact5_alarm":0,"contact6_alarm":0,"relay1":0,"relay2":0,"LED":0,"Temp":0,"Humid":0}

    i2c = board.I2C()
    sensor = adafruit_am2320.AM2320(i2c)

    # Get humidity
    try:
        humid = '{0}'.format(sensor.relative_humidity)
        result.update({"Humid":humid})
    except Exception as e:
        print("Humidity attempt ",attempts,": ",e)
        result.update({"Humid":"NaN"})

    # Get temperature
    try:
        celsius = '{0}'.format(sensor.temperature)
        celsius = float(celsius)
        fahrenheit = round((celsius * 1.8) + 32 ,2)
        result.update({"Temp":fahrenheit})
    except Exception as e:
        print("Temperature attempt ",attempts,": ",e)
        result.update({"Temp":"NaN"})

    # Get the contact, LED and relay status from the device
    # Create a dictionary for values using the BCM numbering
    values_out = {"relay1":23,"relay2":22,"LED":6}
    values_in  = {"contact1_alarm":16,"contact2_alarm":17,"contact3_alarm":24,"contact4_alarm":19,"contact5_alarm":20,"contact6_alarm":21}

    # Set input pins as inputs and put internal resistors into pulldown mode
    for key, input in values_in.items():
        wiringpi.pinMode(input, 0)                        # Set BCM pin to INPUT
        wiringpi.pullUpDnControl(input, 1)                # Put the BCM pin in pull down resistor mode
        result.update({key:wiringpi.digitalRead(input)})  # Read the BCM pin

    # Set output pins as ouputs
    for key, output in values_out.items():
        wiringpi.pinMode(output, 1)                       # Set pin to OUTPUT
        result.update({key:wiringpi.digitalRead(output)}) # Read the BCM pin

    # Print the result for the end user just in case they are running it from the command line
    for key, value in result.items():
        if key == "LED" or key == "Humid" or key == "Temp":
            print(key,"\t:\t",value)
        else:
            print(key,":\t",value)

    # Return the result as a dictionary for ease of use later in the program
    return(result)

def writeDb(result):
    # Write resulting data to the SQLite database on the ramdisk, everything will
    # reference the database so this shouldn't cause any issues with SNMP, etc.
    con = sqlite3.connect('/home/pi/AlarmPiHat/ramdisk/config.db') # Connect to the database
    cur = con.cursor()                         # Init the cursor

    for key, value in result.items():
        # Update contacts
        if 'contact' in key:
            if value == 1:
                sql = "UPDATE config SET %s=datetime('now','localtime') WHERE 1" % (key)
                cur.execute(sql)
            else:
                cur.execute("UPDATE config SET %s=NULL WHERE 1" % key)

        # Update temperature
        if 'Temp' in key:
            sql = "UPDATE config SET temperature='%s' WHERE id=1" % (value)
            cur.execute(sql)

        # Update humidity
        if 'Humid' in key:
            sql = "UPDATE config SET humidity='%s' WHERE id=1" % (value)
            cur.execute(sql)

    con.commit()                               # Commit the changes to the database
    con.close()                                # Close the database connection

def notifier(result):
    # If any value in the results is in an active state, send a notification
    # Base the time between notificaitons on the timer values in the database
    # Check the last notification datetime with current datetime and if x seconds
    # have not passed do not send the notification.
    print("Send notification via email if necessary")
    smtp_server = "mail.totalhighspeed.net"
    port = 587  # For starttls
    sender_email = "chris@totalhighspeed.net"
    receiver_email = "chris@totalhighspeed.net"
    password = input("Enter your password: ")

    # Create a secure SSL context
    context = ssl.create_default_context()

    # Try to log in to server and send email
    try:
        server = smtplib.SMTP(smtp_server,port)
        server.ehlo() # Can be omitted
        server.starttls(context=context) # Secure the connection
        server.ehlo() # Can be omitted
        server.login(sender_email, password)
        # TODO: Send email here
        message = """Subject: TEST FROM AlarmPiHat

This message is a test sent from Python."""
    except Exception as e:
        server.sendmail(sender_email, receiver_email, message)

        # Print any error messages to stdout
        print(e)
    finally:
        server.quit()

#########################
### Main code section ###
#########################
# Run a continuous loop and get the data every x seconds
wiringpi.pinMode(6, 1)     # Set the LED BCM pin to output
wiringpi.digitalWrite(6,1) # Turn on the LED
result = getData()          # Get the data
writeDb(result)             # Write the data to the database
#notifier(result)           # Send notificaiton email if necessary
wiringpi.digitalWrite(6,0)  #Turn off the LED
